backend/cfg/vectorize.ml

```diff
@@ -134,7 +138,327 @@ module Dependency_graph = struct
+      | Load -> args
+      | Store -> Array.sub args 1 (Array.length args - 1)
+
+    let print_memory_chunk ppf (t : t) =
+      let open Format in
+      let str =
+        match t.memory_chunk with
+        | Byte_unsigned -> "Byte_unsigned"
+        | Byte_signed -> "Byte_signed"
+        | Sixteen_unsigned -> "Sixteen_unsigned"
+        | Sixteen_signed -> "Sixteen_signed"
+        | Thirtytwo_unsigned -> "Thirtytwo_unsigned"
+        | Thirtytwo_signed -> "Thirtytwo_signed"
+        | Single _ -> "Single"
+        | Word_int -> "Word_int"
+        | Word_val -> "Word_val"
+        | Double -> "Double"
+        | Onetwentyeight_unaligned -> "Onetwentyeight_unaligned"
+        | Onetwentyeight_aligned -> "Onetwentyeight_aligned"
+      in
+      fprintf ppf "%s" str
+
+    let print_addressing ppf (t : t) =
+      let open Format in
+      let addr = t.addressing_mode in
+      let arg = args t in
+      let printreg ppf (reg : Reg.t) =
+        fprintf ppf "%d,\"%s\"" reg.stamp (Reg.name reg)
+      in
+      match addr with
+      | Ibased (s, _glob, n) -> fprintf ppf "ibased \"%s\" + %i" s n
+      | Iindexed n ->
+        let idx = Printf.sprintf " + %i" n in
+        fprintf ppf "iindexed %a%s" printreg arg.(0) idx
+      | Iindexed2 n ->
+        let idx = Printf.sprintf " + %i" n in
+        fprintf ppf "iindexed2 %a + %a%s" printreg arg.(0) printreg arg.(1) idx
+      | Iscaled (scale, n) ->
+        let idx = Printf.sprintf " + %i" n in
+        fprintf ppf "iscaled %a  * %i%s" printreg arg.(0) scale idx
+      | Iindexed2scaled (scale, n) ->
+        let idx = Printf.sprintf " + %i" n in
+        fprintf ppf "iindexed2scaled %a + %a * %i%s" printreg arg.(0) printreg
+          arg.(1) scale idx
+
+    let dump ppf (t : t) =
+      let open Format in
+      let instruction = t.instruction in
+      fprintf ppf "\nInstruction %d: %a (%a, %a)"
+        (Instruction.id instruction)
+        Cfg.print_instruction instruction print_memory_chunk t print_addressing
+        t
+
+    let compare (t1 : t) (t2 : t) =
+      let addressing_mode_1 = t1.addressing_mode in
+      let addressing_mode_2 = t2.addressing_mode in
+      let args_1 = args t1 in
+      let args_2 = args t2 in
+      let compare_addressing_modes =
+        match addressing_mode_1, addressing_mode_2 with
+        | Ibased (symbol1, global1, n1), Ibased (symbol2, global2, n2) -> (
+          match global1, global2 with
+          | Global, Global | Local, Local ->
+            if symbol1 < symbol2
+            then -1
+            else if symbol1 > symbol2
+            then 1
+            else n1 - n2
+          | Global, Local -> -1
+          | Local, Global -> 1)
+        | Ibased _, _ -> -1
+        | _, Ibased _ -> 1
+        | Iindexed n1, Iindexed n2 ->
+          let arg0_1 = Array.get args_1 0 in
+          let arg0_2 = Array.get args_2 0 in
+          let arg0_compare = arg0_1.stamp - arg0_2.stamp in
+          if arg0_compare = 0 then n1 - n2 else arg0_compare
```

Comment:
The code can be simplified with a helper function `compare_args : Reg.t array -> Reg.t array -> int`.

Reply:
the code now compares it in a different way